oeps/oep-0003.rst

```diff
@@ -0,0 +1,144 @@
+yet.  Clearly the work should be done asynchronously, but we do not yet have a
+standard solution for asynchronous tasks which allow the user to track their
+progress and/or retrieve output upon completion.  Upon further discussion, it
+became clear that other such asynchronous tasks, existing and planned, would
+benefit from a consistent framework.  Examples include:
+
+* Re-running a course
+* Sending bulk email to course participants
+* Generating certificates
+* Grading (rescoring, reset attempts, delete state, calculated grades CSV)
+* Data downloads (grades, student profiles)
+* Open Response Assessment
+* Discussion/forum notifier
+* Uploading and encoding videos (Video Abstraction Layer, VEDA)
+
+
+Rationale
+=========
+
+Celery is already in use for other asynchronous tasks, and made a logical
+choice as the underlying task execution engine.  Course import and export
+timeouts are already happening relatively often, so there is a desire to keep
+the scope of the initial implementation small enough that it can be finished
+fairly quickly.
+
+Similar functionality has already been implemented for
+`course re-runs <https://github.com/edx/edx-platform/tree/master/common/djangoapps/course_action_state>`_
+and assorted LMS actions such as
+`bulk email and grade downloads <https://github.com/edx/edx-platform/tree/master/lms/djangoapps/instructor_task>`_,
+but as one-off implementations that share no code.  The new
+implementation should combine the best features of these in a form that can be
+easily reused for all the user-triggered asynchronous tasks in any service.
+
+The `Jobtastic <https://policystat.github.io/jobtastic/>`_ add-on for Celery
+includes some potentially useful functionality for status tracking and error
+handling, but it's not yet clear if it would be suitable to include as-is for
+this system or just serve as a source of ideas.  There doesn't seem to be a
+good existing library providing any user interface elements for managing
+celery tasks which isn't geared towards system administrators, although some
+of the code in the task dashboard from
+`Flower <https://github.com/mher/flower>`_ may prove useful.
+
+
+Backwards Compatibility
+=======================
+
+* Some documentation that deals with course import and export will need to be
+  updated to describe the new workflow.  The task status page should contain
+  enough information to be pretty self-explanatory, though.
+* There will be migrations to add new tables for task tracking.
+* There are a lot of asynchronous tasks already in the
+  `instructor task <https://github.com/edx/edx-platform/tree/master/lms/djangoapps/instructor_task>`_
+  package of the LMS, so particular care should be taken to produce something
+  compatible with that in both code and behavior.  The existing code is local
+  to the LMS and makes some assumptions about what kind of tasks can be run,
+  so it can't just be used as is.
+
```

Comment:
@jmbowman I'd add a note that we'll likely need to reimplement services over time to fit the new workflow (is that true?), and work with various partners to update their integrations if it's required. I'd stress that this process is opt-in (we'll move things over as we have time) for each service.